gcloud-java-pubsub/src/main/java/com/google/gcloud/pubsub/SubscriptionInfo.java

```diff
@@ -0,0 +1,180 @@
+    private int ackDeadlineSeconds;
+
+    private BuilderImpl(String topic, String name) {
+      this.topic = topic;
+      this.name = name;
+    }
+
+    BuilderImpl(SubscriptionInfo subscription) {
+      name = subscription.name;
+      topic = subscription.topic;
+      pushConfig = subscription.pushConfig;
+      ackDeadlineSeconds = subscription.ackDeadlineSeconds;
+    }
+
+    public Builder name(String name) {
+      this.name = checkNotNull(name);
+      return this;
+    }
+
+    public Builder topic(String topic) {
+      this.topic = checkNotNull(topic);
+      return this;
+    }
+
+    public Builder pushConfig(PushConfig pushConfig) {
+      this.pushConfig = pushConfig;
+      return this;
+    }
+
+    public Builder ackDeadLineSeconds(int ackDeadlineSeconds) {
+      this.ackDeadlineSeconds = ackDeadlineSeconds;
+      return this;
+    }
+
+    public SubscriptionInfo build() {
+      return new SubscriptionInfo(this);
+    }
+  }
+
+  SubscriptionInfo(BuilderImpl builder) {
+    topic = builder.topic;
+    name = builder.name;
+    pushConfig = builder.pushConfig;
+    ackDeadlineSeconds = builder.ackDeadlineSeconds;
+  }
+
+  public String topic() {
+    return topic;
+  }
+
+  public String name() {
+    return name;
+  }
+
+  public PushConfig pushConfig() {
+    return pushConfig;
+  }
+
+  public long ackDeadlineSeconds() {
+    return ackDeadlineSeconds;
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+    return Objects.equals(toPb(), ((SubscriptionInfo)o).toPb());
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(topic, name, pushConfig, ackDeadlineSeconds);
+  }
+
+  @Override
+  public String toString() {
+    return MoreObjects.toStringHelper(this)
+        .add("topic", topic)
+        .add("name", name)
+        .add("pushConfig", pushConfig)
+        .add("ackDeadlineSeconds", ackDeadlineSeconds)
+        .toString();
+  }
+
+  com.google.pubsub.v1.Subscription toPb() {
+    com.google.pubsub.v1.Subscription.Builder builder =
+        com.google.pubsub.v1.Subscription.newBuilder();
+    builder.setTopic(topic);
+    builder.setName(name);
+    builder.setAckDeadlineSeconds(ackDeadlineSeconds);
+    if (pushConfig != null) {
+      builder.setPushConfig(pushConfig.toPb());
+    }
+    return builder.build();
+  }
+
+  static SubscriptionInfo fromPb(com.google.pubsub.v1.Subscription subscription) {
+    Builder builder = builder(subscription.getTopic(), subscription.getName());
+    builder.ackDeadLineSeconds(subscription.getAckDeadlineSeconds());
+    if (subscription.hasPushConfig()) {
+      builder.pushConfig(PushConfig.fromPb(subscription.getPushConfig()));
+    }
+    return builder.build();
+  }
+
+  public Builder toBuilder() {
+    return new BuilderImpl(this);
+  }
+
+  public static SubscriptionInfo of(String topic, String name) {
```

Comment:
So if a push config is not provided the subscription is in pull mode?
Should we provide a `of(String topic, String name, String endpoint)` for fast creation of push subscriptions?